pg_shard.c

```diff
@@ -2007,6 +2008,107 @@ PgShardProcessUtility(Node *parsetree, const char *queryString,
 			}
 		}
 	}
+	else if (statementType == T_DropStmt)
+	{
+		DropStmt *dropStatement = (DropStmt *) parsetree;
+		bool dropBehavior = dropStatement->behavior;
+		bool extensionInstalled = false;
+		Oid extensionOid = InvalidOid;
+		bool missingOK = true;
+
+		extensionOid = get_extension_oid(PG_SHARD_EXTENSION_NAME, missingOK);
+		if (extensionOid != InvalidOid)
+		{
+			extensionInstalled = true;
+		}
+
+		/* handle T_DropStmt for tables and extensions only */
+		if (dropStatement->removeType == OBJECT_TABLE)
+		{
+			ListCell *dropObjectCell = NULL;
+
+			foreach(dropObjectCell, dropStatement->objects)
+			{
+				List *tableNameList = (List *) lfirst(dropObjectCell);
+				RangeVar *rangeVar = makeRangeVarFromNameList(tableNameList);
+				Oid relationId = RangeVarGetRelid(rangeVar, AccessShareLock, true);
+
+				/*
+				 * If a distributed table is dropped with CASCADE modifier, first inform,
+				 * then delete all the dependent metadata information.
+				 */
+				if (extensionInstalled == true && IsDistributedTable(relationId))
+				{
+					if (dropBehavior == DROP_CASCADE)
+					{
+						ereport(NOTICE, (errmsg("removing only metadata for the "
+										        "distributed table"),
+										 errdetail("Shards for this table are not "
+											       "removed.")));
+
+						DeleteDistributedTableMetadata(relationId);
+					}
+					else
+					{
+						char *tableName = get_rel_name(relationId);
+
+						ereport(ERROR, (errcode(ERRCODE_DEPENDENT_OBJECTS_STILL_EXIST),
+										errmsg("cannot drop table %s because other "
+											   "objects depend on it", tableName),
+										errdetail("Table is distributed and metadata "
+												  "information is dependent on it."),
+										errhint("Use DROP ... CASCADE to drop the "
+												"dependent objects too.")));
+					}
+				}
+			}
+		}
+		else if (dropStatement->removeType == OBJECT_EXTENSION)
+		{
+			ListCell *dropStatementObject = NULL;
+
+			foreach(dropStatementObject, dropStatement->objects)
+			{
+				List *objectNameList = lfirst(dropStatementObject);
+				char *objectNameStr = NameListToString(objectNameList);
+
+				if (strcmp(PG_SHARD_EXTENSION_NAME, objectNameStr) == 0)
+				{
+					/*
+					 * If pg_shard is dropped with CASCADE modifier, let it to be dropped.
+					 * However, inform the user that shards on the worker nodes will not
+					 * be dropeed.
+					 *
+					 * If CASCADE modifier is not used, check for existence of any
+					 * distributed tables. If there exists any distributed table, do not
+					 * let pg_shard to be dropped.
+					 */
+					if (extensionInstalled == true && dropBehavior == DROP_CASCADE)
```

Comment:
Also, why do we need to check whether the extension is installed at all? Is it possible for the execution to get this far if it's not (i.e. would our utility hook run before PostgreSQL even verifies whether ever object mentioned in the statement exists?)